### ISSUE TITLE
Handling a django model with m2m fields

### ISSUE THREAD

**Issue by jangeador** [adamghill/django-unicorn#303]
What is the proper way to prevent additional queries in a django model with several m2m relationships. I tried the basic examples and they worked great. When I tried with some of my existing models the unicorn view queried all the models for all the foreign keys resulting in many unnecessary queries. I tried using the `Queryset.values()` method as shown in the example, and that made it work as expected, however, at that point I was passing a dict of values and not really a django model so the filter and other things outlined in the examples did not work as expected. 

I am only using the fields from the `Animal` table. I don't need any of the foreign keys for this component

The view below results in 1 query
```py3
class AnimaltableView(UnicornView):
    animals: QuerySetType[Animal] = Animal.objects.none()

    def mount(self):
        self.load_table()

    def load_table(self):
        self.animals = Animal.objects.all().values('pk', 'name')[0:10]
```

This view results in 61 queries
```py3
class AnimaltableView(UnicornView):
    animals: QuerySetType[Animal] = Animal.objects.none()

    def mount(self):
        self.load_table()

    def load_table(self):
        self.animals = Animal.objects.all()[0:10]
```

This is the template
```html
{% load unicorn %}
<div>
  <table>
    <thead>
    <tr>
      <td>id</td>
      <td>name</td>
    </tr>
    </thead>
    {% for animal in animals %}
      <tr>
        <td>{{ animal.pk }}</td>
        <td>{{ animal.name }}</td>
      </tr>
    {% endfor %}
  </table>
</div>
```

**Comment by adamghill** [comment]
This might be happening when the object gets serialized and it's automatically making those database calls. For serialization I only get ids for FKs and many-to-many fields, though, so I'm surprised. I'll take a look and see if there is anything I can do in unicorn itself to prevent this from happening.

However, you could experiment with either [`defer`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.defer) or [`only`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.only) to only request the fields you actually need.

You might also be able to use [`select_related`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.select_related) or [`prefetch_related`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#prefetch-related) to signal to the ORM that grab more data up-front instead of lazily asking for the data later.

If you are able, can you show me the code for your models? That way I can try to replicate exactly with a test case and see if I can make this better in unicorn itself.

**Comment by jangeador** [comment]
I cannot share the code for the models as the project has information that I am not allowed to share. I will take the time to recreate the models with other values and try to replicate the issue. In the meantime, I changed the queryset to only return one record and these are the results, please look at the comments:

```py3
    def load_table(self):
        # 6 queries
        self.animals = Animal.objects.all()[0:1]
```

```py3
    def load_table(self):
        # 18 queries
        self.animals = Animal.objects.only('tag')[0:1]
```

```py3
    def load_table(self):
        # 6 queries
        self.animals = Animal.objects.select_related()[0:1]
```

```py3
    def load_table(self):
        # 11 queries
        # self.related contains a list of foreign key fields (2)
        # self.lookups contains a list of m2m fields (5)
        self.animals = Animal.objects.select_related(*self.related).prefetch_related(*self.lookups)[0:1]
```

As soon as I have a bit of time, I will provide more information. Thanks for answering so quickly.

**Comment by adamghill** [comment]
I have fix in https://github.com/adamghill/django-unicorn/pull/308 that seems to help. Previously, `prefetch_related` wouldn't be used when serializing the data to JSON, but now it will be.

I have another fix in the works so that `javascript_exclude` would work to prevent the serializer from trying to handle those fields at all, but it's still WIP.

**Comment by adamghill** [comment]
This has been merged and released in 0.39.1. Let me know what your results are with the new code!
